File: java/P03.java
// P03 Find the Kth element of a list.

import core.List;

import static core.List.*;

class P03 {
    // Recursion
    <T> T f1(List<T> list, int n) {
        if (list.isEmpty()) throw new IndexOutOfBoundsException();
        if (n < 1) throw new IndexOutOfBoundsException();
        return n == 1 ? list.head() : f1(list.tail(), n - 1);
    }

    // Calls chain
    <T> T f2(List<T> list, int n) {
        return list.drop(n - 1).head();
    }

    // Using index
    <T> T f3(List<T> list, int n) {
        return list.get(n - 1);
    }
}

File: scala/P03.scala
// P03 Find the Kth element of a list.

// Tail recursion
@annotation.tailrec
def f1[T](list: List[T], n: Int): T = list match {
  case Nil => throw new IndexOutOfBoundsException()
  case h :: t =>
    if (n < 1) throw new IndexOutOfBoundsException()
    if (n == 1) h else f1(t, n - 1)
}

// Calls chain
def f2[T](list: List[T], n: Int): T = list.drop(n - 1).head

// Using index
def f3[T](list: List[T], n: Int): T = list(n - 1)

// With error reporting
def f4[T](list: List[T], n: Int): T = list.lift(n - 1).getOrElse(throw new IndexOutOfBoundsException())